landoapi/transplants.py

```diff
@@ -301,6 +304,56 @@ def warning_wip_commit_message(*, revision, **kwargs):
         return "This revision is marked as a WIP. Please remove `WIP:` before landing."
 
 
+@RevisionWarningCheck(8, "Repository is under a soft code freeze.", True)
+def warning_code_freeze(*, repo, **kwargs):
+    supported_repos = get_repos_for_env(current_app.config.get("ENVIRONMENT"))
+    try:
+        repo_details = supported_repos[repo["fields"]["shortName"]]
+    except KeyError:
+        return
+
+    if repo_details.product_details_url:
+        default_message = "Could not retrieve repository's code freeze status."
+
+        try:
+            product_details = requests.get(repo_details.product_details_url).json()
+        except requests.exceptions.RequestException as e:
+            logger.exception(e)
+            return [{"message": default_message}]
+
+        # If the JSON doesn't have these keys, this warning isn't applicable
+        try:
+            freeze_date_str = product_details["NEXT_SOFTFREEZE_DATE"]
+            merge_date_str = product_details["NEXT_MERGE_DATE"]
+        except KeyError:
+            return
+
+        # The code freeze dates generally correspond to PST work days.
+        utc_offset = "-0800"
+        today = datetime.now(tz=timezone.utc)
+        freeze_date = datetime.strptime(
+            f"{freeze_date_str} {utc_offset}",
+            "%Y-%m-%d %z",
+        ).replace(tzinfo=timezone.utc)
+        if today < freeze_date:
+            return
+
+        merge_date = datetime.strptime(
+            f"{merge_date_str} {utc_offset}",
+            "%Y-%m-%d %z",
+        ).replace(tzinfo=timezone.utc)
+
+        if freeze_date <= today <= merge_date:
+            return [
+                {
+                    "message": (
+                        f"Repository is under a soft code freeze "
+                        f"(ends {merge_date_str})."
+                    )
+                }
+            ]
```

Comment:
slight cleanup here to reduce nesting and make sure not to return an incorrect message if a repo does not implement the product details URL:

```suggestion
    if not repo_details.product_details_url:
        # Repo does not have a product details URL.
        return

    try:
        product_details = requests.get(repo_details.product_details_url).json()
    except requests.exceptions.RequestException as e:
        logger.exception(e)
        return [{"message": "Could not retrieve repository's code freeze status."}]

    # If the JSON doesn't have these keys, this warning isn't applicable
    try:
        freeze_date_str = product_details["NEXT_SOFTFREEZE_DATE"]
        merge_date_str = product_details["NEXT_MERGE_DATE"]
    except KeyError:
        return

    # The code freeze dates generally correspond to PST work days.
    utc_offset = "-0800"
    today = datetime.now(tz=timezone.utc)
    freeze_date = datetime.strptime(
        f"{freeze_date_str} {utc_offset}",
        "%Y-%m-%d %z",
    ).replace(tzinfo=timezone.utc)
    if today < freeze_date:
        return

    merge_date = datetime.strptime(
        f"{merge_date_str} {utc_offset}",
        "%Y-%m-%d %z",
    ).replace(tzinfo=timezone.utc)

    if freeze_date <= today <= merge_date:
        return [
            {
                "message": (
                    f"Repository is under a soft code freeze "
                    f"(ends {merge_date_str})."
                )
            }
        ]                                                                                                                                                                                  

```